src/core/sandbox.c

```diff
@@ -0,0 +1,801 @@
+/*
+ sandbox.c : irssi
+
+    Copyright (C) 2015 Namsun Ch'o
+
+    This program is free software; you can redistribute it and/or modify
+    it under the terms of the GNU General Public License as published by
+    the Free Software Foundation; either version 2 of the License, or
+    (at your option) any later version.
+
+    This program is distributed in the hope that it will be useful,
+    but WITHOUT ANY WARRANTY; without even the implied warranty of
+    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+    GNU General Public License for more details.
+
+    You should have received a copy of the GNU General Public License along
+    with this program; if not, write to the Free Software Foundation, Inc.,
+    51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
+*/
+
+#include <errno.h>
+#include <linux/futex.h>
+#include <linux/in.h>
+#include <linux/prctl.h>
+#include <seccomp.h>
+#include <signal.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <sys/capability.h>
+#include <sys/fcntl.h>
+#include <sys/ioctl.h>
+#include <sys/mman.h>
+#include <sys/prctl.h>
+#include <sys/resource.h>
+#include <sys/socket.h>
+#include <sys/termios.h>
+#include <sys/wait.h>
+#include <time.h>
+#include <unistd.h>
+
+#include "sandbox.h"
+
+#define IRSSI_NEEDS_RDTSC
+#define IRSSI_NEEDS_CAPS
+
+/*
+ * TODO:
+ *	- implement user, pid, network, and mount namespaces
+ *	- ensure that hardcoded file descriptors work
+ *	- create automated test suite for seccomp (and/or static analysis?)
+ *	- keep private keys in a separate process, (e.g https://github.com/AGWA/titus)?
+ *	- add support tame() on OpenBSD, and Capsicum on FreeBSD
+*/
+
+void create_namespaces(void)
+{
+	/* TODO */
+}
+
+void drop_privileges(void)
+{
+	int tsc_state;
+	size_t i;
+	cap_t cap;
+	cap_flag_value_t value;
+
+	/* running as root largely defeats the purpose of using a sandbox */
+	if (geteuid() == 0 || getegid() == 0) {
+		fprintf(stderr, "Cannot run irssi as root when the sandbox is enabled\n");
+		exit(1);
+	}
+
+	/* disable the rdtsc so it cannot be used for side-channel attacks */
+#ifndef IRSSI_NEEDS_RDTSC
+	if (prctl(PR_GET_TSC, &tsc_state, 0, 0, 0) < 0) {
+		fprintf(stderr, "Could not get the TSC state\n");
+		exit(1);
+	}
+
+	if (tsc_state == PR_TSC_ENABLE) {
+		prctl(PR_SET_TSC, PR_TSC_SIGSEGV, 0, 0, 0);
+		prctl(PR_GET_TSC, &tsc_state, 0, 0, 0);
+
+		if (tsc_state != PR_TSC_SIGSEGV) {
+			fprintf(stderr, "Could not disable the TSC\n");
+			exit(1);
+		}
+	}
+#endif
+
+	/* drop any dangerous capabilities the process is running with
+	 * https://forums.grsecurity.net/viewtopic.php?f=7&t=2522 */
+	/* FIXME why does cap_get_proc() fail? */
+#ifndef IRSSI_NEEDS_CAPS
+	cap = cap_get_proc();
+	if (cap == NULL) {
+		fprintf(stderr, "Could not initialize capabilities (cap_get_proc failed)\n");
+		exit(1);
+	}
+
+	if (cap_clear(cap) < 0) {
+		fprintf(stderr, "Could not clear capabilities (cap_clear failed)\n");
+		cap_free(cap);
```

Comment:
@ahf: gotcha. but is there any reason to clean up heap variables before `exit()`?